### PR TITLE
Add Etag header

### LINKED ISSUE #191: Implement ETag
ETags aren't my favorite caching mechanism because all the heavy lifting is done by the server, but it might be worth looking into doing them anyway. Do we do strong ETags or weak ones? Should this be customizable from the Caddyfile? What are the default settings? How do we cache the computed ETags?

The Go standard library supports ETag and If-Modified-Since headers, but we need to do our part to take advantage of them.

### PR DESCRIPTION
**Pull request by jupiter**
This is on par with Nginx, using last modified and size. Yes, it is not bullet-proof, but it is an Etag header, an optimization that should not cost more than it saves. Fixes #191

### CODE REVIEW

**Comment by mholt** [comment]
Hey @jupiter, thanks for the PR! This is looking great.

I'm going to see if @weingart will take a look at this in the next few days as I know he's been working on some caching features too.

File: middleware/fileserver.go [hunk 1]
```diff
@@ -128,6 +129,10 @@ func (fh *fileHandler) serveFile(w http.ResponseWriter, r *http.Request, name st
 		return http.StatusNotFound, nil
 	}
 
+	// Add ETag header
+	e := fmt.Sprintf("\"%x-%x\"", d.ModTime().Unix(), d.Size())
```
Comment (weingart) [inline middleware/fileserver.go]: It's a small thing, but I really don't like exposing more information than necessary.  While the last modification time may also be returned by the http request, putting it in here seems "wrong".

However, that can be changed later on.  :)

File: middleware/fileserver_test.go [hunk 2]
```diff
@@ -60,12 +62,14 @@ func TestServeHTTP(t *testing.T) {
 			url:                 "https://foo/file1.html",
 			expectedStatus:      http.StatusOK,
 			expectedBodyContent: testFiles["file1.html"],
+			expectedEtag:        "\"1e240-13\"",
```
Comment (weingart) [inline middleware/fileserver_test.go]: Using `` strings might make this easier to read.

File: middleware/fileserver_test.go [hunk 3]
```diff
@@ -197,6 +210,12 @@ func beforeServeHTTPTest(t *testing.T) {
 			return
 		}
 		f.Close()
+
+		// and set the last modified time
+		err = os.Chtimes(absFile, fixedTime, fixedTime)
```
Comment (weingart) [inline middleware/fileserver_test.go]: This will (should!) fail if/when this test is not run as root.

Comment (weingart) [inline middleware/fileserver_test.go]: A better option (likely) would be to structure a test case on something similar to "failingFS", where the returned modification time is constant.

File: middleware/fileserver.go [hunk 4]
```diff
@@ -128,6 +129,10 @@ func (fh *fileHandler) serveFile(w http.ResponseWriter, r *http.Request, name st
...
+	w.Header().Set("Etag", e)
```
Comment (weingart) [inline middleware/fileserver.go]: Please change this to "ETag", a la:

https://en.wikipedia.org/wiki/HTTP_ETag
https://tools.ietf.org/html/rfc7232#section-2.3

Yes, headers are case insensitive, but we should be strict in what we send, and liberal in what we accept.  :)

**Comment by weingart** [comment]
Other than the test and the "Etag" header, I think this can go in.  I'll re-structure my code around this.  :)

File: middleware/fileserver_test.go (hunk 3 above)
Comment (jupiter) [inline middleware/fileserver_test.go]: I don't run tests as root, but this works. Are you sure?

**Comment by mholt** [comment]
Alrighty, here we go then! Thanks for the review. And thank you very much @jupiter for the pull request!

**Comment by mholt** [comment]
Bah. I forgot to squash. Curse these new GitHub features :smile:

**Comment by jupiter** [comment]
Damn, I was going to, but wanted to see if there was anything else first.

**Comment by mholt** [comment]
@jupiter It's okay. I've made you a collaborator, so you can help review pull requests, create new ones on branches in this repository, and help close issues. Again, thanks for your help!

**Comment by jupiter** [comment]
Thank you, glad to be able to help.

**Comment by MarkHerhold** [comment]
Does this PR mean that the Etag header is automatically added in future Caddy releases or does it have to be enabled in the config file? If it has to be enabled/configured, what option is that?

Sorry for the dumb questions - I'm new to Caddy.

**Comment by mholt** [comment]
@MarkHerhold The ETag header is automatically added for static files. This will be available at the next Caddy release.

**Comment by coolaj86** [comment]
Is there any documentation for this?

Etags are enabled by default for `file_server`?

**Comment by mohammed90** [comment]
> Etags are enabled by default for `file_server`?

Yes, see:
https://github.com/caddyserver/caddy/blob/0e2c7e1d35b287fc0e56d6db2951f791e09b5a37/modules/caddyhttp/fileserver/staticfiles.go#L431

**Pull request merged**